Replace projects conn assign with parameter

mbramson opened the pull request:
The PairRetro :new and :edit actions were assigning the projects used
for the dropdown to the conn rather than simply passing through the
projects to the view.

This also fixes a bug where if any errors occurred creating or updating
a PairRetro, the projects dropdown would not populate.

Refactor PairRetroController and the PairRetro model.
  - Add pair and project parameters to the changesets.
  - Remove earliest_pair_date paramter from changeset (since this information is on the pair that is now passed in).
  - Form for retros now only lists projects that are associated with the retro through the retro's pair's group.

Closes #58

Closes #84 

Further Issues
- [x] Disallow :edit and :update actions if the user is no longer in the group associated with the retro's pair. Also make the edit button the PairRetro :index not appear for retros where this is the case.

mbramson committed:
Replace projects conn assign with parameter

The PairRetro :new and :edit actions were assigning the projects used
for the dropdown to the conn rather than simply passing through the
projects to the view.

This also fixes a bug where if any errors occurred creating or updating
a PairRetro, the projects dropdown would not populate.

mbramson committed:
Add pair_with_users_and_group/1

mbramson committed:
Refactor pair_retro model tests

mbramson committed:
Remove redundant pair_with_users function head

add tests for pair_with_users/1

mbramson committed:
Add projects_for_group/1

mbramson committed:
Add project associated with retro validation

Make the retro changeset validate that the passed in project is
associated with the group that the pair that the retro is associated
with is associated with (what a sentence)

Make the pair_retro new form only list projects that the pair is
associated with through that pair's group.

Make the pair_retro :create action validate that the passed in project
is associated with the pair through the group.

mbramson committed:
Make retro create return proper projects on error

mbramson committed:
Make retro :edit only list group's projects

mbramson committed:
Validate projects in retro update for group

mbramson committed:
Remove pair_date from retro changesets

This information is already contained in the pair that is passed to the
changeset.

mbramson committed:
Change order of retro changeset parameters

mbramson committed:
Refactor PairRetroController

ericworkman commented on web/models/pair.ex:
```diff
@@ -26,12 +26,18 @@ defmodule Pairmotron.Pair do
     preload: [users: users]
   end
+
+  def pair_with_users_and_group(pair_id) do
```
Is this solely used for testing?

mbramson replied:
Oooooh. This is vestigial. I forgot to remove it.

ericworkman commented on web/controllers/pair_retro_controller.ex:
```diff
@@ -67,35 +70,42 @@ defmodule Pairmotron.PairRetroController do
   end
 
+  defp project_from_params_or_pair_retro(params, pair_retro) do
+    id = Map.get(params, "project_id") || (pair_retro.project && pair_retro.project.id) || 0
```
The `|| 0` is here so that the `Repo.get` works right? I think it would be better to return `nil` instead of making a database call and relying on there never being a Project 0.

mbramson replied:
Ya, that's exactly why. The trouble is `Repo.get(Project, nil)` errors. There really should never be a project 0, as the index starts at 0 and counts up. There might be a better way to handle this.

ericworkman commented on web/models/pair_retro.ex:
```diff
@@ -48,15 +52,20 @@ defmodule Pairmotron.PairRetro do
+  defp validate_date_is_not_before_pair(changeset, field, pair) do
+    pair_start_date = cond do
+      is_nil(pair) -> nil
+      true -> Pairmotron.Calendar.first_date_of_week(pair.year, pair.week)
```
This could be two functions if you pattern match on pair == nil. Looks like it might simplify logic below in the changeset validation.

mbramson replied:
Definitely! Good call.

ericworkman reviewed (commented):
Looks good, just a couple comments.

mbramson committed:
Remove vestigial pair_with_users_and_groups

Make pair_with_users use a left_join so that it returns a pair in the
case that there are no users in that pair.

mbramson committed:
Simplify validate_date_is_not_before_pair/3

mbramson committed:
Refactor project_from_params_or_pair_retro/2

mbramson commented:
@ericworkman addressed your concerns, let me know what you think!

ericworkman reviewed (approved):
LGTM!
